Bug 1321798 - Rewrite autoclassification frontend for better sheriff workflow

jgraham opened the pull request:
Please ignore the first two commits in this review (i.e. any changes to non-js files); I need to rebase and land that work. But the js changes are rather large and it's way past time to get some review on them, so I think it's worth starting this process anyway.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/mozilla/treeherder/2089)
<!-- Reviewable:end -->

jgraham commented:
Ah, also the fourth commit.

wlach commented on ui/plugins/auto_classification/controller.js:
```diff
@@ -69,891 +40,1104 @@ treeherder.factory('thStringOverlap', function() {
+                $scope.optionChanged();
             }
+            if (option === "=") {
+                $("#" + line.id + "-manual-bug").focus();
```
Kind of a bummer to be using raw jquery here, but I assume you couldn't find an alternative.

wlach commented on ui/plugins/auto_classification/controller.js:
```diff
@@ -69,891 +40,1104 @@ treeherder.factory('thStringOverlap', function() {
-                    }
+            var resources = {
+                "matchers": ThMatcherModel.by_id(),
+               
```
Is this useful?

wlach replied:
Nit: Do you actually need/want to create a temporary variable here?

wlach commented on ui/plugins/auto_classification/controller.js:
```diff
@@ -69,891 +40,1104 @@ treeherder.factory('thStringOverlap', function() {
+         */
+        function build() {
+            log.debug("build", ctrl);
+            log.debug("Status", ctrl.loadStatus);
```
Did you want to leave this in?

camd commented on ui/plugins/auto_classification/controller.js:
```diff
@@ -69,891 +40,1104 @@ treeherder.factory('thStringOverlap', function() {
+         * @param {Object[]} extraOptions - List of line-independent options
+         * @param {Object[]} prevLine - Line before the current one in the log
+         */
+        function getDefaultOption(options, extraOptions, prevLine) {
```
This function would really benefit from more comments.  You have some comments here, which helps, but the logic is pretty complex, so it'd be MUCH easier to review and understand later if you put in a play-by-play to some degree.

Something along the lines of: "Now we have determined that X is true, so we need to find out if we should Y or Z."  I may be the only one, but comments like that help me with a decision function such as this.  :)

jgraham commented:
Review status: 0 of 49 files reviewed at latest revision, 17 unresolved discussions.

---

*[ui/index.html, line 141 at r12](https://reviewable.io:443/reviews/mozilla/treeherder/2089#-Kam33RJEp2aFrZ8xX3k-r12-141:-KccAPlDQ5n39BKXrW98:b-896fix) ([raw file](https://github.com/mozilla/treeherder/blob/5dd7edd6fe83ec93d16d772bc12c98465fce7b47/ui/index.html#L141)):*
<details><summary><i>Previously, camd wrote…</i></summary><blockquote>

won't this allow the config file to be added to the minified js?  I don't think we want that, do we?  :)
</blockquote></details>

Done.

---

*[ui/js/controllers/bugfiler.js, line 10 at r12](https://reviewable.io:443/reviews/mozilla/treeherder/2089#-Kam33RPxfxjnUCWQHI2-r12-10:-KccAQxrIvd-4Kk0KDkq:b-896fix) ([raw file](https://github.com/mozilla/treeherder/blob/5dd7edd6fe83ec93d16d772bc12c98465fce7b47/ui/js/controllers/bugfiler.js#L10)):*
<details><summary><i>Previously, wlach (William Lachance) wrote…</i></summary><blockquote>

It seems like there is a bunch of cleanup in here which isn't strictly related to your feature work. I'd prefer if this was put into another PR.
</blockquote></details>

Done.

---

*[ui/js/controllers/main.js, line 344 at r12](https://reviewable.io:443/reviews/mozilla/treeherder/2089#-Kam33RJEp2aFrZ8xX3l-r12-344:-KcPp8ZU0QEswpRmIVyg:b-4lri5w) ([raw file](https://github.com/mozilla/treeherder/blob/5dd7edd6fe83ec93d16d772bc12c98465fce7b47/ui/js/controllers/main.js#L344)):*
<details><summary><i>Previously, camd wrote…</i></summary><blockquote>

this would be good to send out a note that these shortcuts will be changing.  I don't have a feel for how many people may use ``j`` and ``k`` this way, but it seems ripe for breaking someone's workflow...  :)  Doesn't mean we shouldn't do it.  Pretty unnecessary to have 2 keys to do the same thing...
</blockquote></details>

Sure, I have mentioned this to the sheriffs several times,  but I can mention it again with the inevitable launch email.

---

*[ui/js/controllers/main.js, line 525 at r12](https://reviewable.io:443/reviews/mozilla/treeherder/2089#-Kam33RJEp2aFrZ8xX3l-r12-525:-Kcc7q1uWktQcVmhJkCY:b-896fix) ([raw file](https://github.com/mozilla/treeherder/blob/5dd7edd6fe83ec93d16d772bc12c98465fce7b47/ui/js/controllers/main.js#L525)):*
<details><summary><i>Previously, camd wrote…</i></summary><blockquote>

would you go ahead and comment the rest of these?  It will save someone some time down the line... :)
</blockquote></details>

Done.

---

*[ui/js/models/matcher.js, line 36 at r12](https://reviewable.io:443/reviews/mozilla/treeherder/2089#-Kam33RKg-BDFn85PESz-r12-36:-Kcc7tDop36zZqchD_ir:b-896fix) ([raw file](https://github.com/mozilla/treeherder/blob/5dd7edd6fe83ec93d16d772bc12c98465fce7b47/ui/js/models/matcher.js#L36)):*
<details><summary><i>Previously, camd wrote…</i></summary><blockquote>

you don't actually need the parens around ``response`` and ``elem`` since it's just one param each
</blockquote></details>

Done.

---

*[ui/js/models/text_log_errors.js, line 28 at r12](https://reviewable.io:443/reviews/mozilla/treeherder/2089#-Kam33RKg-BDFn85PET7-r12-28:-Kcc7vONAt0e6Yhbb8yD:b-896fix) ([raw file](https://github.com/mozilla/treeherder/blob/5dd7edd6fe83ec93d16d772bc12c98465fce7b47/ui/js/models/text_log_errors.js#L28)):*
<details><summary><i>Previously, wlach (William Lachance) wrote…</i></summary><blockquote>

Couldn't you just use `map` here?
</blockquote></details>

Done.

---

*[ui/plugins/tabs.js, line 8 at r12](https://reviewable.io:443/reviews/mozilla/treeherder/2089#-Kam33RSAS3uf1TusH_0-r12-8:-KccA01V85raQw3ooIZL:b-896fix) ([raw file](https://github.com/mozilla/treeherder/blob/5dd7edd6fe83ec93d16d772bc12c98465fce7b47/ui/plugins/tabs.js#L8)):*
<details><summary><i>Previously, wlach (William Lachance) wrote…</i></summary><blockquote>

Likewise, could you put these changes in a separate PR?
</blockquote></details>

Done.

---

*[ui/plugins/auto_classification/controller.js, line 71 at r12](https://reviewable.io:443/reviews/mozilla/treeherder/2089#-Kam33RKg-BDFn85PET1-r12-71:-KccA3iM6a_fbbo8y6XM:b-896fix) ([raw file](https://github.com/mozilla/treeherder/blob/5dd7edd6fe83ec93d16d772bc12c98465fce7b47/ui/plugins/auto_classification/controller.js#L71)):*
<details><summary><i>Previously, camd wrote…</i></summary><blockquote>

You could just ``return`` here since you don't use the var anywhere.
</blockquote></details>

Done.

---

*[ui/plugins/auto_classification/controller.js, line 117 at r12](https://reviewable.io:443/reviews/mozilla/treeherder/2089#-Kam33RKg-BDFn85PET1-r12-117:-KccA5O5yxq624_OMuwD:b-896fix) ([raw file](https://github.com/mozilla/treeherder/blob/5dd7edd6fe83ec93d16d772bc12c98465fce7b47/ui/plugins/auto_classification/controller.js#L117)):*
<details><summary><i>Previously, camd wrote…</i></summary><blockquote>

Looks like a few of these bindings are never used.  Could they be removed?
</blockquote></details>

Done.

---

*[ui/plugins/auto_classification/controller.js, line 167 at r12](https://reviewable.io:443/reviews/mozilla/treeherder/2089#-Kam33RKg-BDFn85PET1-r12-167:-KccA7ae93jZss8pyzq-:b-exae48) ([raw file](https://github.com/mozilla/treeherder/blob/5dd7edd6fe83ec93d16d772bc12c98465fce7b47/ui/plugins/auto_classification/controller.js#L167)):*
<details><summary><i>Previously, wlach (William Lachance) wrote…</i></summary><blockquote>

If you're going to parametrize the bugzilla link, may as well use that value here too.
</blockquote></details>

Done in the other PR; I'll rebase on to that when it lands.

---

*[ui/plugins/auto_classification/controller.js, line 442 at r12](https://reviewable.io:443/reviews/mozilla/treeherder/2089#-Kam33RKg-BDFn85PET1-r12-442:-KccAS2KfZxfEvECkvL2:b-896fix) ([raw file](https://github.com/mozilla/treeherder/blob/5dd7edd6fe83ec93d16d772bc12c98465fce7b47/ui/plugins/auto_classification/controller.js#L442)):*
<details><summary><i>Previously, camd wrote…</i></summary><blockquote>

typo: "of if" sb "or if"
</blockquote></details>

Done.

---

*[ui/plugins/auto_classification/controller.js, line 496 at r12](https://reviewable.io:443/reviews/mozilla/treeherder/2089#-Kam33RKg-BDFn85PET1-r12-496:-KccAEVygZPZi0Ta85-y:b-896fix) ([raw file](https://github.com/mozilla/treeherder/blob/5dd7edd6fe83ec93d16d772bc12c98465fce7b47/ui/plugins/auto_classification/controller.js#L496)):*
<details><summary><i>Previously, camd wrote…</i></summary><blockquote>

This function would really benefit from more comments.  You have some comments here, which helps, but the logic is pretty complex, so it'd be MUCH easier to review and understand later if you put in a play-by-play to some degree.

Something along the lines of: "Now we have determined that X is true, so we need to find out if we should Y or Z."  I may be the only one, but comments like that help me with a decision function such as this.  :)
</blockquote></details>

Done.

---

*[ui/plugins/auto_classification/controller.js, line 507 at r12](https://reviewable.io:443/reviews/mozilla/treeherder/2089#-Kam33RKg-BDFn85PET1-r12-507:-KccAIZNAH5wk923-PAs:b-896fix) ([raw file](https://github.com/mozilla/treeherder/blob/5dd7edd6fe83ec93d16d772bc12c98465fce7b47/ui/plugins/auto_classification/controller.js#L507)):*
<details><summary><i>Previously, wlach (William Lachance) wrote…</i></summary><blockquote>

It's not that big a deal, but you could shorten this to:

```js
var parts = line.split(" | ", 3);
return (parts.length === 3) parts[1] : null;
```
</blockquote></details>

Done.

---

*[ui/plugins/auto_classification/controller.js, line 606 at r12](https://reviewable.io:443/reviews/mozilla/treeherder/2089#-Kam33RKg-BDFn85PET1-r12-606:-KcPukRA0Q1Bt0DXWyas:b-9xqpnd) ([raw file](https://github.com/mozilla/treeherder/blob/5dd7edd6fe83ec93d16d772bc12c98465fce7b47/ui/plugins/auto_classification/controller.js#L606)):*
<details><summary><i>Previously, wlach (William Lachance) wrote…</i></summary><blockquote>

Kind of a bummer to be using raw jquery here, but I assume you couldn't find an alternative.
</blockquote></details>

I don't really know how to programmatically focus an element using angular. Unless it's easy it doesn't seem worthwhile.

---

*[ui/plugins/auto_classification/controller.js, line 837 at r12](https://reviewable.io:443/reviews/mozilla/treeherder/2089#-Kam33RKg-BDFn85PET1-r12-837:-KccAKNgssp3SshqdYSA:bss4a2b) ([raw file](https://github.com/mozilla/treeherder/blob/5dd7edd6fe83ec93d16d772bc12c98465fce7b47/ui/plugins/auto_classification/controller.js#L837)):*
<details><summary><i>Previously, wlach (William Lachance) wrote…</i></summary><blockquote>

Did you want to leave this in?
</blockquote></details>

Removed.

---

*[ui/plugins/auto_classification/controller.js, line 907 at r12](https://reviewable.io:443/reviews/mozilla/treeherder/2089#-Kam33RKg-BDFn85PET1-r12-907:-KccAO9Q80df_fpB9W8h:b-896fix) ([raw file](https://github.com/mozilla/treeherder/blob/5dd7edd6fe83ec93d16d772bc12c98465fce7b47/ui/plugins/auto_classification/controller.js#L907)):*
<details><summary><i>Previously, wlach (William Lachance) wrote…</i></summary><blockquote>

Nit: Do you actually need/want to create a temporary variable here?
</blockquote></details>

Done.

---

*[Gruntfile.js, line 167 at r12](https://reviewable.io:443/reviews/mozilla/treeherder/2089#-Kam33RJEp2aFrZ8xX3d-r12-167:-KcPoQe21ZRj_a1PbiJ6:b-n7o3za) ([raw file](https://github.com/mozilla/treeherder/blob/5dd7edd6fe83ec93d16d772bc12c98465fce7b47/Gruntfile.js#L167)):*
<details><summary><i>Previously, wlach (William Lachance) wrote…</i></summary><blockquote>

Was this change intentional?
</blockquote></details>

Well yes, but I can revert if needed. It just makes it easier to debug the minified code by adding whitespace. Presumably there is some small cost in load time.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/treeherder/2089)*
<!-- Sent from Reviewable.io -->

wlach commented on tests/autoclassify/test_classify_failures.py:
```diff
@@ -37,45 +32,54 @@ def do_autoclassify(job, test_failure_lines, matchers, status="testfailed"):
-        assert [item.id for item in actual.classified_failures.all()] == [expected.id]
+    for (error_line, failure_line), expected in zip(zip(*expected_classified),
+                                                    classified_failures):
+        assert [item.id for item in error_line.classified_failures.all()] == [expected.id]
```
Not that it matters hugely, but I would do:

```py
error_line.classified_failures.values_list('id', flat=True) == [expected.id]
```

wlach commented on tests/autoclassify/test_classify_failures.py:
```diff
@@ -37,45 +32,54 @@ def do_autoclassify(job, test_failure_lines, matchers, status="testfailed"):
-def test_classify_test_failure(failure_lines, classified_failures,
+def test_classify_test_failure(text_log_errors_failure_lines,
+                               classified_failures,
                                test_job_2):
```
Could you add a short comment stating what cases this test is supposed to be testing? Some of the other test cases in this file are obvious from the name, but`test_classify_test_failure` is pretty ambiguous...

wlach commented on tests/autoclassify/test_classify_failures.py:
```diff
@@ -37,45 +32,54 @@ def do_autoclassify(job, test_failure_lines, matchers, status="testfailed"):
-                                               (test_line, {"message": "message2"})])
+def test_no_autoclassify_job_success(text_log_errors_failure_lines,
+                                     classified_failures,
+                                     test_job_2):
```
Ditto, a short comment describing what's being tested here would be helpful.

wlach commented on tests/ui/unit/controllers/autoclassify.tests.js:
```diff
@@ -1,104 +1,104 @@
 'use strict';
 
-describe('AutoClassify panel', function() {
+// describe('AutoClassify panel', function() {
```
I'll leave it up to you on whether it is worth refactoring this to reflect the current state of the code. But if you're just going to comment this out, I would rather remove the file.

jgraham replied:
Removed it for now. I think tests here probably are useful though.

wlach commented on treeherder/model/migrations/0055_autoclassify_status.py:
```diff
@@ -0,0 +1,38 @@
+SET FOREIGN_KEY_CHECKS=0;
+ALTER TABLE `job` ADD COLUMN `autoclassify_status` integer DEFAULT 0 NOT NULL;
+ALTER TABLE `job` ALTER COLUMN `autoclassify_status` DROP DEFAULT;
+COMMIT;
```
I am curious if the manual SQL for this migration has any advantages over the automatically generated sql by Django? I know (boy do I know) that sometimes Django can be inefficient at doing these things, but I don't see how that would be the case here.

I'd prefer to use the default behaviour where possible, as it would make an eventual migration to another database easier (at the very least there would be less to audit).

jgraham replied:
Only if the `set foreign_key_checks` thing is a good idea in this case. Maybe it's not since no foreign keys are affected.

edmorley replied:
In cases where we override the default migrations, I think we should include the original sqlmigrate output in a comment in the migration file.

I'll check the migrations more thoroughly when this is added.

wlach commented on treeherder/model/models.py:
```diff
@@ -1450,15 +1497,120 @@ class TextLogError(models.Model):
+    best_classification = models.ForeignKey("ClassifiedFailure",
+                                            related_name="best_for_errors",
+                                            null=True,
+                                            db_index=True,
```
db_index is not necessary here, as django automatically indexes by column where a foreign key is specified.

wlach commented on treeherder/webapp/api/text_log_error.py:
```diff
@@ -0,0 +1,122 @@
+        for item in data:
+            line_id = int(item.get("id"))
+            if line_id is None:
+                return "No text log error id provided", HTTP_400_BAD_REQUEST
```
if `item.get(..)` is None, casting it to an int is going to throw an exception.

wlach commented on treeherder/model/models.py:
```diff
@@ -1441,6 +1471,23 @@ class Meta:
                            'finished_line_number')
 
 
+class TextLogErrorManager(models.Manager):
```
Could you add one-line comments for each of the methods in here describing what they do?

wlach commented on treeherder/autoclassify/autoclassify.py:
```diff
@@ -20,57 +26,85 @@ def match_errors(job):
+        matches, all_matched = find_matches(unmatched_errors)
+        update_db(job, matches, all_matched)
+    except:
+        logger.debug("Autoclassification of job %s failed" % job.id)
```
Shouldn't this be log level error?

wlach commented on treeherder/autoclassify/detect_intermittents.py:
```diff
@@ -1,8 +1,8 @@
 import logging
```
Given that we aren't running this, I feel like it might be better just to remove it rather than continuing to maintain code that we're not really testing.

jgraham replied:
Well we are still running the unit tests for it.

wlach replied:
Yes but that's not at all the same as code we're running in production. I'm not going to block review on this, but IMO you should just remove these files as they're apt to confuse people. If it turns out that we want to resurrect this code, it's easy to go back in history.

jgraham replied:
I don't think it's "different" to the code we're running in production; it's just preffed off. That seems fine to me. Nevertheless, I removed it, but it at least seems pretty weird to bundle it with these unrelated changes.

wlach replied:
You could have opened another pull request for the removal, I wasn't saying that it had to happen as part of this PR.

edmorley commented on ui/plugins/failure_summary/main.html:
```diff
@@ -5,7 +5,6 @@
             <div class="job-tabs-content">
                 <a class="btn btn-xs btn-default"
                    prevent-default-on-left-click
-                   ng-show="user.loggedin && (filerInAddress || user.is_staff)"
```
Did you mean to make this change?

edmorley commented on ui/plugins/controller.js:
```diff
@@ -56,7 +56,8 @@ treeherder.controller('PluginCtrl', [
         var showAutoClassifyTab = function() {
             thTabs.tabs.autoClassification.enabled = ($location.search().autoclassify === true ||
-                                                      $location.host().indexOf('herokuapp.com') !== -1) &&
+                                                      $location.host().indexOf('herokuapp.com') !== -1 ||
```
This could be in a separate PR.

edmorley commented on ui/partials/main/thShortcutTable.html:
```diff
@@ -14,56 +14,86 @@
-        <td><span class="kbd">ctrl</span><span class="kbd">shift
-          </span><span class="kbd">f</span></td>
-        <td>Clear the quick filter</td>
+        <th colspan=2>Jobs View</th>
```
Much of the changes in this file could be in a separate PR.

edmorley commented on ui/js/controllers/main.js:
```diff
@@ -341,14 +341,14 @@ treeherderApp.controller('MainCtrl', [
 
             // Shortcut: select next unclassified failure
-            [['j', 'n'], function() {
+            ['n', function() {
```
I can't speak for what the current sheriffs use, but I only ever used j/k and not n/p, so this change would have really messed with my muscle memory. Though happy for them to make the call on this.

edmorley commented on ui/js/controllers/bugfiler.js:
```diff
@@ -2,12 +2,14 @@
-        thPinboard, thEvents, fullLog, parsedLog, reftest, selectedJob, allFailures,
-        thNotify) {
+        fullLog, parsedLog, reftest, selectedJob, allFailures,
+        successCallback, thNotify) {
```
Yeah please can this be in a separate PR.

edmorley commented on ui/css/treeherder-global.css:
```diff
@@ -333,9 +333,11 @@ input:focus::-moz-placeholder {
 #onscreen-shortcuts {
-    position: relative;
-    top: 130px;
+    position: fixed;
```
Separate PR please!

edmorley commented on treeherder/model/migrations/0055_autoclassify_status.py:
```diff
@@ -0,0 +1,38 @@
+
+    operations = [migrations.RunSQL(
+        sql="""
+BEGIN;
```
Would you mind using the list form of RunSQL rather than a string, since the latter has to be parsed by sqlparse, which I'd just managed to remove our use of (I need to update the Django docs to make the list form more obvious).

eg:
https://github.com/mozilla/treeherder/blob/5318924e00157ac8dc76351d527a970a4f6db26f/treeherder/model/migrations/0001_squashed_0053_add_job_platform_option_push_index.py#L725-L728

jgraham replied:
OK, but why? This is a pretty substantial extra burden since there is no longer a string in the migration file that you can copy and check for syntax errors / run in a SQL command line. I think that the more requirements we place on people trying to use migrations for invasive changes, the more likely they are to decide to run the SQL manually against production instead.

edmorley replied:
Ok fair point, happy to leave as is.

edmorley commented on treeherder/log_parser/tasks.py:
```diff
@@ -88,8 +90,19 @@ def crossreference_error_lines(job_id, priority):
+        has_lines = crossreference_job(job)
+    except Exception as e:
+        import traceback
+        logger.error(traceback.format_exception(e))
```
Wouldn't `logger.exception()` give the same result, but without the need to use `traceback` manually?

That said, do we need this? We don't do any such special handling anywhere else.

edmorley commented on treeherder/log_parser/tasks.py:
```diff
@@ -64,6 +64,7 @@ def parse_log(job_log, priority):
     Call ArtifactBuilderCollection on the given job.
     """
     post_log_artifacts(job_log)
+    logger.debug("Scheduling crossreference for job %i from parse_log" % job_log.job.id)
```
This could be in a separate PR.

edmorley commented on treeherder/log_parser/crossreference.py:
```diff
@@ -63,6 +76,8 @@ def _crossreference(job):
                 line_number=error.line_number,
                 failure_line=failure_line))
+            error.failure_line = failure_line
+            error.save()
```
This looks like it could be in a separate PR?

edmorley commented on treeherder/autoclassify/management/commands/detect_intermittents.py:
```diff
@@ -1,15 +0,0 @@
-from treeherder.model.models import Job
-
-
-class Command(BaseCommand):
```
Could we remove the detect intermittents code in a separate, earlier PR?

edmorley commented:
Just spotted this New Relic exception on treeherder-prototype:
https://rpm.newrelic.com/accounts/677903/applications/7385291/filterable_errors?tw%5Bend%5D=1487030691&tw%5Bstart%5D=1487028891#/show/bdac3cb6-f248-11e6-9c7c-b82a72d22a14_0_4582/stack_trace?top_facet=transactionUiName&primary_facet=error.class&barchart=barchart&_k=5hdtqb

KeyError during the autoclassify task (autoclassify.py line 84, in update_db).

edmorley commented:
Just thought - it's probably worth testing the "ES isn't present" codepaths still work (since on initial landing the ES instances won't exist or be populated for stage/prod). On prototype I'm presuming this could be tested by clearing the `ELASTICSEARCH_URL` environment variable temporarily.

On a similar note, do we think we need to test the "ES is present but is empty" case on prototype too? Seeing as we'll presumably need to run the backfill script on stage/prod so should test it works still.

wlach commented on ui/js/models/job_artifact.js:
```diff
@@ -0,0 +1,45 @@
+'use strict';
+
+treeherder.factory('ThJobArtifactModel', [
```
LOL no, let's not add this monster back. ;)

wlach commented on ui/plugins/auto_classification/main.html:
```diff
@@ -1,181 +1,10 @@
+   autoclassify-status="job.autoclassify_status"
+   th-user="user"
+   >
+</th-autoclasify-panel>
```
```html
</th-autoclassify-panel>
```
